drivers/network/netio/netio.c

```diff
@@ -0,0 +1,860 @@
+    {
+        goto err_out;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        goto err_out_free_nc;
+    }
+
+	/* TODO: @thomasfabber: how do I undo this in case the
+         * TdiSendDatagram routine fails? Or does it handle
+         * that gracefully?
+         */
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        goto err_out_free_nc_and_tci;
+    }
+
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+
+    /* This will create a tdiIrp: */
+    status = TdiSendDatagram(&tdiIrp, s->LocalAddressFile, ((char *)BufferData) + Buffer->Offset,
+                        Buffer->Length, TargetConnectionInfo, NetioComplete, nc);
+
+    /* If allocating tdiIrp fails we get here.
+     * Call the IoCompletion of the application's Irp so this Irp
+     * gets freed.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        goto err_out_free_nc_and_tci_unmap;
+    }
+    return STATUS_PENDING;
+
+err_out_free_nc_and_tci_unmap:
+    /* TODO: implement freeing Buffer mmap */
+
+err_out_free_nc_and_tci:
+    ExFreePool(TargetConnectionInfo);
+
+err_out_free_nc:
+    ExFreePoolWithTag(nc, 'NEIO');
+
+err_out:
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return STATUS_PENDING;
+}
+
+static NTSTATUS WSKAPI
+WskReceiveFrom(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _Out_opt_ PSOCKADDR RemoteAddress,
+    _Inout_ PULONG ControlLength,
+    _Out_writes_bytes_opt_(*ControlLength) PCMSGHDR ControlInfo,
+    _Out_opt_ PULONG ControlFlags, _Inout_ PIRP Irp)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static NTSTATUS WSKAPI
+WskReleaseUdp(_In_ PWSK_SOCKET Socket, _In_ PWSK_DATAGRAM_INDICATION DatagramIndication)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static NTSTATUS WSKAPI
+WskReleaseTcp(_In_ PWSK_SOCKET Socket, _In_ PWSK_DATA_INDICATION DataIndication)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static NTSTATUS WSKAPI
+WskGetLocalAddress(_In_ PWSK_SOCKET Socket, _Out_ PSOCKADDR LocalAddress, _Inout_ PIRP Irp)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static NTSTATUS WSKAPI
+WskGetRemoteAddress(_In_ PWSK_SOCKET Socket, _Out_ PSOCKADDR RemoteAddress, _Inout_ PIRP Irp)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static NTSTATUS WSKAPI
+WskSocketConnect(
+    PWSK_CLIENT Client,
+    USHORT SocketType,
+    ULONG Protocol,
+    PSOCKADDR LocalAddress,
+    PSOCKADDR RemoteAddress,
+    ULONG Flags,
+    PVOID SocketContext,
+    const WSK_CLIENT_CONNECTION_DISPATCH * Dispatch,
+    PEPROCESS OwningProcess,
+    PETHREAD OwningThread,
+    PSECURITY_DESCRIPTOR SecurityDescriptor,
+    PIRP Irp)
```

Comment:
Please add the missing SAL2 annotations.

Reply:
Ok done. Also added annotations for other functions.